cpp/src/arrow/compute/kernels/aggregate_basic_internal.h

```diff
@@ -891,6 +891,105 @@ struct NullMinMaxImpl : public ScalarAggregator {
   }
 };
 
+template <SimdLevel::type SimdLevel>
+struct DictionaryMinMaxImpl : public ScalarAggregator {
+  using ThisType = DictionaryMinMaxImpl<SimdLevel>;
+
+  DictionaryMinMaxImpl(std::shared_ptr<DataType> out_type, ScalarAggregateOptions options)
+      : options(std::move(options)),
+        out_type(std::move(out_type)),
+        has_nulls(false),
+        count(0),
+        min(nullptr),
+        max(nullptr) {
+    this->options.min_count = std::max<uint32_t>(1, this->options.min_count);
+  }
+
+  Status Consume(KernelContext*, const ExecSpan& batch) override {
+    if (batch[0].is_scalar()) {
+      return Status::NotImplemented("No min/max implemented for DictionaryScalar");
+    }
+
+    DictionaryArray arr(batch[0].array.ToArrayData());
+    this->has_nulls = arr.null_count() > 0;
+    this->count += arr.length() - arr.null_count();
+
+    Datum dict_values(arr.dictionary());
+    ARROW_ASSIGN_OR_RAISE(Datum result, MinMax(std::move(dict_values)));
```

Comment:
Would it be possible that a value doesn't exist in any dict value?

Reply:
Sure, I add a if statement to avoid redundant calculations.

Reply:
https://arrow.apache.org/docs/format/Columnar.html#dictionary-encoded-layout

I'm not so familiar with dictionary, would that be possible?

```
dictionary: ["1", "2", "3"]
indices: [0, 1, 0, 1]
```

Reply:
> ```
> dictionary: ["1", "2", "3"]
> indices: [0, 1, 0, 1]
> ```

It's possible.  When a user constructs a DictionaryArray via specifying its dictionary and indices, it could happen.
To handle it, we need to validate all the indexes in dictionary could be found in indices at least once. I wonder whether we already have the validation codes.

Reply:
I'm not sure what you mean with "validate", but the example you gave above is perfectly valid.

Reply:
You'll probably want to "Filter" rather than "Validate". You can use 'is_in' and 'filter' to do such filtration.

Reply:
Hmm, perhaps a more efficient routine can be written specifically for this (find all dictionary indices actually used in a dictionary array)? Going through "is_in" and "filter" sounds quite expensive.

Reply:
It could probably also be useful for other kernels.

Reply:
Would "is_in" or filter consume more memory than expected? Since previous MinMax should be `O(1)` in space and `O(array-length)` for computing. And "is_in" for dictionary indices would take `O(dictionary-length)` when accepting a dictionary. I don't know would it matter?

Reply:
I agree that `is_in` isn't ideal in this scenario. However, I don't think `O(1)` space is possible, because we need to somehow save which indices to keep in the dictionary?

Reply:
Hmmm I think here we only need to keep the scalar?

Reply:
Yes if we don't want to reuse the primitive minmax kernels. But if we do, we'll have to compact the dictionary and then pass it to the kernels.

Reply:
> Interestingly, the same feature was just requested for the go implementation #37277.

I see @zeroshade have taken this. I think we could wait for this issue to be fixed and handle it with the dictionay compaction.

Reply:
I think in some scenerio, like user decode a huge dictionary for string when reading from underlying data( like mentioned in https://github.com/apache/arrow/pull/35825 ), making a dictionary array would be a disaster...

But any implemention look good to me :-)

Reply:
> https://github.com/apache/arrow/blob/main/cpp/src/parquet/column_writer.cc#L1683-L1688
> 
> (I just found that parquet uses a "Take" here)

We have two ways to implement the MinMax Function For a DicionaryArray.

1. Take/DictionaryDecode firstly and then reuse MinMax Function with the input of the decoded Array.
2. DictionaryCompaction firstly and then reuse MinMax Function with the input of the compacted dictionary.

Each Time Complexity is O(n), I am not sure which implementation is better.

Reply:
> I think we should still insist on the latter one. Since, for DicionaryArray whose dictionary is already compacted, the latter implementation could be faster a little.

Yeah I think it's ok, just move forward!

Reply:
What I'm afraid is a corner case, that when user is using BinaryLike dictionary, and if it's large, a lots of binary would be copied. However just move forward, I think it's just a corner case and we can support different mode later.